# <font color=blue>Assignments for "Evaluating Goodness of Fit"</font>

As in previous lessons, please submit a link to a single gist that contains links to two Juypyter notebooks (one for each assignment below).

## 1. Weather model

For this assignment, you'll revisit the historical temperature dataset. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.preprocessing import StandardScaler
from sklearn.preprocessing import Normalizer

%matplotlib inline
weather_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/weatherHistory.csv")
weather_df.head()

/usr/local/lib/python3.7/dist-packages/statsmodels/tools/_testing.py:19: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


Formatted Date  ...                      Daily Summary
0  2006-04-01 00:00:00.000 +0200  ...  Partly cloudy throughout the day.
1  2006-04-01 01:00:00.000 +0200  ...  Partly cloudy throughout the day.
2  2006-04-01 02:00:00.000 +0200  ...  Partly cloudy throughout the day.
3  2006-04-01 03:00:00.000 +0200  ...  Partly cloudy throughout the day.
4  2006-04-01 04:00:00.000 +0200  ...  Partly cloudy throughout the day.

[5 rows x 12 columns]

- Like in the previous lesson, build a linear regression model where your target variable is the difference between the *apparenttemperature* and the *temperature*. As explanatory variables, use *humidity* and *windspeed*. Now, estimate your model using OLS. What are the R-squared and adjusted R-squared values? Do you think they are satisfactory? Why? 

In [ ]:
y = weather_df['Temperature (C)'] - weather_df['Apparent Temperature (C)']

X = weather_df[['Humidity', 'Wind Speed (km/h)']]

In [ ]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Mon, 31 May 2021   Prob (F-statistic):               0.00
Time:                        11:58:41   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


Both R-squared and adjusted R-squared are equal and 0.288. Generally, situations where the R-square value is 0.7 and above are desirable. In particular, we can say that R-square values ​​below 0.3 are actually not good.

- Next, include the interaction of *humidity* and *windspeed* to the model above and estimate the model using OLS. Now, what is the R-squared of this model? Does this model improve upon the previous one? 


In [ ]:
X['humidity*windspeed'] = X['Humidity'] * X['Wind Speed (km/h)']

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Mon, 31 May 2021   Prob (F-statistic):               0.00
Time:                        11:58:41   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity*windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""


Now the new R-squared values ​​are 0.341. improved compared to the previous model. It is slightly higher than the threshold value of 0.3.

- Add *visibility* as additional explanatory variable to the first model and estimate it. Did R-squared increase? What about adjusted R-squared? Compare the differences put on the table by the interaction term and the *visibility* in terms of the improvement in the adjusted R-squared. Which one is more useful?

In [ ]:
X['visibility'] = weather_df['Visibility (km)']
X.drop('humidity*windspeed',axis=1, inplace=True)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Mon, 31 May 2021   Prob (F-statistic):               0.00
Time:                        11:58:42   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
visibility           -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Although both R-squared and adjusted R-squared have increased by comparison with the first model, the interaction term resulted in a better (adjusted) R-squared in the previous model.

- Choose the best one from the three models above with respect to their AIC and BIC scores. Validate your choice by discussing your justification with your mentor.

For both AIC and BIC, the lower value the better. Therefore, the second model, that is, the model with the interaction term was better with respect to their AIC and BIC scores.

##  2. House prices model

In this exercise, you'll work on your house prices model. To complete this assignment, submit a link in the gist file to the Jupyter notebook containing your solutions to the following tasks:

In [ ]:
house_prices_df = pd.read_csv("https://djl-lms-assets.s3.eu-central-1.amazonaws.com/datasets/house_prices.csv", sep = ";")
house_prices_df.head()

Id  MSSubClass MSZoning  ...  SaleType  SaleCondition SalePrice
0   1          60       RL  ...        WD         Normal    208500
1   2          20       RL  ...        WD         Normal    181500
2   3          60       RL  ...        WD         Normal    223500
3   4          70       RL  ...        WD        Abnorml    140000
4   5          60       RL  ...        WD         Normal    250000

[5 rows x 81 columns]

- Run your house prices model again and assess the goodness of fit of your model using F-test, R-squared, adjusted R-squared, AIC and BIC.

In [ ]:
df_null = house_prices_df.isna().sum() / len(house_prices_df)
house_prices_df.drop(df_null[df_null > .1].index, axis=1, inplace=True)
house_prices_df.dropna(inplace=True)


In [ ]:
cols=['OverallQual', 'GrLivArea', 'GarageCars', 'TotalBsmtSF', 'FullBath', 'YearBuilt','YearRemodAdd']
df_train = house_prices_df[cols]

MSZoning = pd.get_dummies(house_prices_df['MSZoning'], drop_first=True)
Street = pd.get_dummies(house_prices_df['Street'], drop_first=True)
BsmtQual = pd.get_dummies(house_prices_df['BsmtQual'], drop_first=True)
SaleCondition = pd.get_dummies(house_prices_df['SaleCondition'], drop_first=True)

scaler = StandardScaler()

df_train = pd.DataFrame(scaler.fit_transform(df_train), columns=df_train.columns, index=df_train.index)

X = pd.concat([df_train, MSZoning, Street, BsmtQual,SaleCondition],axis =1)

y = house_prices_df['SalePrice']

In [ ]:
X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.792
Model:                            OLS   Adj. R-squared:                  0.788
Method:                 Least Squares   F-statistic:                     250.0
Date:                Mon, 31 May 2021   Prob (F-statistic):               0.00
Time:                        11:58:42   Log-Likelihood:                -15937.
No. Observations:                1338   AIC:                         3.192e+04
Df Residuals:                    1317   BIC:                         3.202e+04
Df Model:                          20                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const         2.016e+05   1.95e+04     10.330      0.000    1.63e+05     2.4e+05
OverallQual   2.362e+04   1690.376     13.975      0.000    2.03e+04    2.69e+04
GrLivArea     2.777e+04   1580.325     17.570      0.000    2.47e+04    3.09e+04
GarageCars    9197.9477   1393.891      6.599      0.000    6463.458    1.19e+04
TotalBsmtSF   8882.9745   1297.002      6.849      0.000    6338.558    1.14e+04
FullBath     -3428.4919   1524.637     -2.249      0.025   -6419.475    -437.509
YearBuilt     2346.1049   1871.396      1.254      0.210   -1325.138    6017.348
YearRemodAdd  6115.1872   1382.217      4.424      0.000    3403.599    8826.775
FV            2.316e+04   1.48e+04      1.560      0.119   -5963.123    5.23e+04
RH            1.748e+04   1.76e+04      0.994      0.320    -1.7e+04     5.2e+04
RL            2.933e+04   1.39e+04      2.107      0.035    2023.981    5.66e+04
RM            1.015e+04    1.4e+04      0.726      0.468   -1.73e+04    3.76e+04
Pave         -7421.8280   1.75e+04     -0.425      0.671   -4.17e+04    2.68e+04
Fa            -4.87e+04   8901.960     -5.471      0.000   -6.62e+04   -3.12e+04
Gd            -4.49e+04   4271.186    -10.513      0.000   -5.33e+04   -3.65e+04
TA           -4.598e+04   5421.943     -8.481      0.000   -5.66e+04   -3.53e+04
AdjLand       2.165e+04   3.66e+04      0.592      0.554   -5.01e+04    9.34e+04
Alloca        1.247e+04   1.47e+04      0.850      0.395   -1.63e+04    4.12e+04
Family       -6742.6003   9073.887     -0.743      0.458   -2.45e+04    1.11e+04
Normal        7992.6754   4139.062      1.931      0.054    -127.200    1.61e+04
Partial       1.575e+04   5600.916      2.812      0.005    4762.575    2.67e+04
==============================================================================
Omnibus:                      638.917   Durbin-Watson:                   1.966
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            81591.508
Skew:                          -1.189   Prob(JB):                         0.00
Kurtosis:                      41.182   Cond. No.                         74.4
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

* F-statistic: 250
* R-squared: 0.792
* Adjusted R-squared: 0.788
* AIC: 31920.0
* BIC: 32020.0


- Do you think your model is satisfactory? If so, why?


The model has considerably R-squared and adjusted R-squared and F-statistic values, so it can be satisfactory.

- In order to improve the goodness of fit of your model, try different model specifications by adding or removing some variables. 

In [ ]:
#cols=['OverallQual', 'GrLivArea', 'GarageCars', 'TotalBsmtSF', 'FullBath', 'YearBuilt','YearRemodAdd']

df_train = house_prices_df.select_dtypes(exclude='O').drop('SalePrice', axis=1)

MSZoning = pd.get_dummies(house_prices_df['MSZoning'], drop_first=True)
Street = pd.get_dummies(house_prices_df['Street'], drop_first=True)
BsmtQual = pd.get_dummies(house_prices_df['BsmtQual'], drop_first=True)
SaleCondition = pd.get_dummies(house_prices_df['SaleCondition'], drop_first=True)

scaler = StandardScaler()

df_train = pd.DataFrame(scaler.fit_transform(df_train), columns=df_train.columns, index=df_train.index)

X = pd.concat([df_train, MSZoning, Street, BsmtQual,SaleCondition],axis =1)

y = house_prices_df['SalePrice']

X = sm.add_constant(X)

results = sm.OLS(y, X).fit()

results.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.827
Model:                            OLS   Adj. R-squared:                  0.821
Method:                 Least Squares   F-statistic:                     131.3
Date:                Mon, 31 May 2021   Prob (F-statistic):               0.00
Time:                        11:58:43   Log-Likelihood:                -15811.
No. Observations:                1338   AIC:                         3.172e+04
Df Residuals:                    1290   BIC:                         3.197e+04
Df Model:                          47                                         
Covariance Type:            nonrobust                                         
=================================================================================
                    coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------
const          1.861e+05   1.87e+04      9.970      0.000    1.49e+05    2.23e+05
Id              115.6065    930.573      0.124      0.901   -1709.996    1941.209
MSSubClass    -5655.6122   1180.945     -4.789      0.000   -7972.395   -3338.830
LotArea        4541.0966   1044.966      4.346      0.000    2491.077    6591.116
OverallQual    2.072e+04   1677.261     12.351      0.000    1.74e+04     2.4e+04
OverallCond    6931.1390   1199.860      5.777      0.000    4577.248    9285.030
YearBuilt      8665.8395   2427.460      3.570      0.000    3903.638    1.34e+04
YearRemodAdd   1726.4090   1496.454      1.154      0.249   -1209.341    4662.159
MasVnrArea     3949.6750   1103.355      3.580      0.000    1785.108    6114.242
BsmtFinSF1     3535.2114   1073.894      3.292      0.001    1428.441    5641.981
BsmtFinSF2      397.2268    952.954      0.417      0.677   -1472.283    2266.736
BsmtUnfSF      -220.5487   1020.660     -0.216      0.829   -2222.884    1781.786
TotalBsmtSF    3922.8159   1725.264      2.274      0.023     538.185    7307.447
1stFlrSF       6899.8421   1831.918      3.766      0.000    3305.978    1.05e+04
2ndFlrSF        1.06e+04   1360.523      7.794      0.000    7935.391    1.33e+04
LowQualFinSF    188.1281    960.710      0.196      0.845   -1696.598    2072.854
GrLivArea      1.411e+04   1447.135      9.750      0.000    1.13e+04    1.69e+04
BsmtFullBath   4535.4601   1374.222      3.300      0.001    1839.505    7231.416
BsmtHalfBath    275.4053    989.813      0.278      0.781   -1666.414    2217.225
FullBath       1494.1691   1647.265      0.907      0.365   -1737.444    4725.782
HalfBath      -1723.6158   1390.817     -1.239      0.215   -4452.127    1004.895
BedroomAbvGr  -5412.7987   1445.896     -3.744      0.000   -8249.365   -2576.233
KitchenAbvGr  -4166.4786   1133.911     -3.674      0.000   -6390.991   -1941.966
TotRmsAbvGrd   5438.4709   2022.521      2.689      0.007    1470.680    9406.261
Fireplaces     2679.7556   1156.815      2.316      0.021     410.311    4949.200
GarageYrBlt   -1292.6118   1943.554     -0.665      0.506   -5105.485    2520.262
GarageCars     9729.7750   1903.202      5.112      0.000    5996.064    1.35e+04
GarageArea     -378.4885   1906.309     -0.199      0.843   -4118.295    3361.318
WoodDeckSF     3161.9161   1016.646      3.110      0.002    1167.454    5156.378
OpenPorchSF   -1275.4245   1038.495     -1.228      0.220   -3312.748     761.899
EnclosedPorch  1209.7623   1044.108      1.159      0.247    -838.574    3258.099
3SsnPorch       904.2633    926.393      0.976      0.329    -913.139    2721.666
ScreenPorch    3391.7537    968.400      3.502      0.000    1491.942    5291.565
PoolArea      -1741.6692    985.700     -1.767      0.077   -3675.420     192.081
MiscVal        -230.7312    926.670     -0.249   

- For each model you try, get the goodness of fit metrics and compare your models with each other. Which model is the best and why?

BEFORE
- R-squared:	0.792
- Adj. R-squared:	0.788
- F-statistic:	250.0
- AIC:	31920.0
- BIC:	32020.0


AFTER
-	R-squared:	0.827
- Adj. R-squared:	0.821
- F-statistic:	131.3
- AIC:	31720.0
- BIC:	31970.0
	


In the first model, we used numeric features with a correlation higher than the value of 0.6 that you previously specified with our target variable, but in the second model we included all numeric values. As a result, we can say that the second model gives better results.